Tests/RxSwiftTests/Tests/Observable+SingleTest.swift

```diff
@@ -558,6 +558,213 @@ extension ObservableSingleTest {
+            ])
+
+        var numberOfTimesInvoked = 0
+
+        let res = scheduler.start { xs.doOnNext { error in
+                numberOfTimesInvoked = numberOfTimesInvoked + 1
+            }
+        }
+
+        let correctMessages = [
+            next(210, 2),
+            next(220, 3),
+            next(230, 4),
+            next(240, 5),
+            completed(250)
+        ]
+
+        let correctSubscriptions = [
+            Subscription(200, 250)
+        ]
+
+        XCTAssertEqual(res.events, correctMessages)
+        XCTAssertEqual(xs.subscriptions, correctSubscriptions)
+
+        XCTAssertEqual(numberOfTimesInvoked, 4)
+    }
+
+    func testDoOnNext_throws() {
+        let scheduler = TestScheduler(initialClock: 0)
+
+        let xs = scheduler.createHotObservable([
+            next(150, 1),
+            next(210, 2),
+            next(220, 3),
+            next(230, 4),
+            next(240, 5),
+            completed(250)
+            ])
+
+        var numberOfTimesInvoked = 0
+
+        let res = scheduler.start { xs.doOnNext { error in
+                if numberOfTimesInvoked > 2 {
+                    throw testError
+                }
+                numberOfTimesInvoked = numberOfTimesInvoked + 1
+            }
+        }
+
+        let correctMessages = [
+            next(210, 2),
+            next(220, 3),
+            next(230, 4),
+            error(240, testError)
+        ]
+
+        let correctSubscriptions = [
+            Subscription(200, 240)
+        ]
+
+        XCTAssertEqual(res.events, correctMessages)
+        XCTAssertEqual(xs.subscriptions, correctSubscriptions)
+
+        XCTAssertEqual(numberOfTimesInvoked, 3)
+    }
+
+    func testDoOnError_normal() {
+        let scheduler = TestScheduler(initialClock: 0)
+
+        let xs = scheduler.createHotObservable([
+            next(150, 1),
+            next(210, 2),
+            error(250, testError)
+            ])
+
+        var recordedError: ErrorType!
+        var numberOfTimesInvoked = 0
+
+        let res = scheduler.start { xs.doOnError { error in
+                recordedError = error
+                numberOfTimesInvoked = numberOfTimesInvoked + 1
+            }
+        }
+
+        let correctMessages = [
+            next(210, 2),
+            error(250, testError)
+        ]
+
+        let correctSubscriptions = [
+            Subscription(200, 250)
+        ]
+
+        XCTAssertEqual(res.events, correctMessages)
+        XCTAssertEqual(xs.subscriptions, correctSubscriptions)
+
+        XCTAssertEqual(recordedError as NSError, testError)
+        XCTAssertEqual(numberOfTimesInvoked, 1)
+    }
+
+    func testDoOnError_throws() {
+        let scheduler = TestScheduler(initialClock: 0)
+
+        let xs = scheduler.createHotObservable([
+            next(150, 1),
+            next(210, 2),
+            error(250, testError)
+            ])
+
+        let res = scheduler.start { xs.doOnError { _ in
+                throw testError
```

Comment:
Could you maybe throw `testError1` here and then have
```
let correctMessages = [
 +            next(210, 2),
 +            error(250, testError1)
 +        ]
```

... so we know that error thrown in `doOnError` overrides original error.